### ISSUE TITLE
ld.lld: error: undefined symbol: vtable for mlir::LLVM::LLVMDialect

### ISSUE THREAD

**Issue by oroppas** [flang-compiler/f18-llvm-project#1067]
While building shared libraries with debug:

```
[0/260] Performing build step for 'libcxx_fuzzer_x86_64'
ninja: no work to do.
[16/260] Linking CXX shared library lib/libFIRTransforms.so.13git
FAILED: lib/libFIRTransforms.so.13git
: && /home/ryuta/local/bin/clang++ -fPIC -stdlib=libc++ -fPIC -fvisibility-inlines-hidden -Werror=date-time -Werror=unguarded-availability-new -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers -pedantic -Wno-long-long -Wc++98-compat-extra-semi -Wimplicit-fallthrough -Wcovered-switch-default -Wno-noexcept-type -Wnon-virtual-dtor -Wdelete-non-virtual-dtor -Wsuggest-override -Wstring-conversion -Wmisleading-indentation -fdiagnostics-color -flto=thin -Wno-deprecated-copy -Wno-string-conversion -Wno-unused-command-line-argument -Wstring-conversion           -Wcovered-switch-default -Wno-nested-anon-types -g -DDEBUGF18  -stdlib=libc++ -Wl,-z,defs -Wl,-z,nodelete -fuse-ld=lld -Wl,--color-diagnostics -flto=thin -Wl,--thinlto-cache-dir=/home/ryuta/packages/llvm/flang/build/lto.cache -shared -Wl,-soname,libFIRTransforms.so.13git -o lib/libFIRTransforms.so.13git tools/flang/lib/Optimizer/Transforms/CMakeFiles/obj.FIRTransforms.dir/AbstractResult.cpp.o tools/flang/lib/Optimizer/Transforms/CMakeFiles/obj.FIRTransforms.dir/AffineDemotion.cpp.o tools/flang/lib/Optimizer/Transforms/CMakeFiles/obj.FIRTransforms.dir/AffinePromotion.cpp.o tools/flang/lib/Optimizer/Transforms/CMakeFiles/obj.FIRTransforms.dir/ArrayValueCopy.cpp.o tools/flang/lib/Optimizer/Transforms/CMakeFiles/obj.FIRTransforms.dir/CharacterConversion.cpp.o tools/flang/lib/Optimizer/Transforms/CMakeFiles/obj.FIRTransforms.dir/ControlFlowConverter.cpp.o tools/flang/lib/Optimizer/Transforms/CMakeFiles/obj.FIRTransforms.dir/CSE.cpp.o tools/flang/lib/Optimizer/Transforms/CMakeFiles/obj.FIRTransforms.dir/ExternalNameConversion.cpp.o tools/flang/lib/Optimizer/Transforms/CMakeFiles/obj.FIRTransforms.dir/FirLoopResultOpt.cpp.o tools/flang/lib/Optimizer/Transforms/CMakeFiles/obj.FIRTransforms.dir/MemDataFlowOpt.cpp.o tools/flang/lib/Optimizer/Transforms/CMakeFiles/obj.FIRTransforms.dir/MemToReg.cpp.o tools/flang/lib/Optimizer/Transforms/CMakeFiles/obj.FIRTransforms.dir/RewriteLoop.cpp.o  -Wl,-rpath,/home/ryuta/packages/llvm/flang/build/lib  lib/libFIRAnalysis.so.13git  lib/libFIRBuilder.so.13git  lib/libFIRDialect.so.13git  lib/libMLIRAffineToStandard.so.13git  lib/libFIRSupport.so.13git  lib/libMLIRTransforms.so.13git  lib/libMLIRCopyOpInterface.so.13git  lib/libMLIRVector.so.13git  lib/libMLIRAffineUtils.so.13git  lib/libMLIRTransformUtils.so.13git  lib/libMLIRLoopAnalysis.so.13git  lib/libMLIRPresburger.so.13git  lib/libMLIRRewrite.so.13git  lib/libMLIRPDLToPDLInterp.so.13git  lib/libMLIRPass.so.13git  lib/libMLIRAnalysis.so.13git  lib/libMLIRSCF.so.13git  lib/libMLIRLinalg.so.13git  lib/libMLIRAffine.so.13git  lib/libMLIRLoopLikeInterface.so.13git  lib/libMLIRMemRef.so.13git  lib/libMLIRDialect.so.13git  lib/libMLIRMemRefUtils.so.13git  lib/libMLIRStandard.so.13git  lib/libMLIRCallInterfaces.so.13git  lib/libMLIRControlFlowInterfaces.so.13git  lib/libMLIRVectorInterfaces.so.13git  lib/libMLIRTensor.so.13git  lib/libMLIRCastInterfaces.so.13git  lib/libMLIRViewLikeInterface.so.13git  lib/libMLIRDialectUtils.so.13git  lib/libMLIRParser.so.13git  lib/libMLIRPDLInterp.so.13git  lib/libMLIRPDL.so.13git  lib/libMLIRSideEffectInterfaces.so.13git  lib/libMLIRInferTypeOpInterface.so.13git  lib/libMLIRDataLayoutInterfaces.so.13git  lib/libMLIRIR.so.13git  lib/libMLIRSupport.so.13git  -lpthread  lib/libLLVMCore.so.13git  lib/libLLVMSupport.so.13git  -Wl,-rpath-link,/home/ryuta/packages/llvm/flang/build/lib && :
ld.lld: error: undefined symbol: vtable for mlir::LLVM::LLVMDialect
>>> referenced by LLVMOpsDialect.h.inc:15 (tools/mlir/include/mlir/Dialect/LLVMIR/LLVMOpsDialect.h.inc:15)
>>>               /home/ryuta/packages/llvm/flang/build/lto.cache/llvmcache-DAE4595E943590B9DC7B515EA98CFB19A090FA80:(mlir::LLVM::LLVMDialect::LLVMDialect(mlir::MLIRContext*))
>>> the vtable symbol may be undefined because the class is missing its key function (see https://lld.llvm.org/missingkeyfunction)

ld.lld: error: undefined symbol: mlir::LLVM::LLVMDialect::initialize()
>>> referenced by LLVMOpsDialect.h.inc:17 (tools/mlir/include/mlir/Dialect/LLVMIR/LLVMOpsDialect.h.inc:17)
>>>               /home/ryuta/packages/llvm/flang/build/lto.cache/llvmcache-DAE4595E943590B9DC7B515EA98CFB19A090FA80:(mlir::LLVM::LLVMDialect::LLVMDialect(mlir::MLIRContext*))

ld.lld: error: undefined symbol: mlir::acc::OpenACCDialect::initialize()
>>> referenced by OpenACCOpsDialect.h.inc:17 (tools/mlir/include/mlir/Dialect/OpenACC/OpenACCOpsDialect.h.inc:17)
>>>               /home/ryuta/packages/llvm/flang/build/lto.cache/llvmcache-DAE4595E943590B9DC7B515EA98CFB19A090FA80:(mlir::acc::OpenACCDialect::OpenACCDialect(mlir::MLIRContext*))

ld.lld: error: undefined symbol: mlir::omp::OpenMPDialect::initialize()
>>> referenced by OpenMPOpsDialect.h.inc:19 (tools/mlir/include/mlir/Dialect/OpenMP/OpenMPOpsDialect.h.inc:19)
>>>               /home/ryuta/packages/llvm/flang/build/lto.cache/llvmcache-DAE4595E943590B9DC7B515EA98CFB19A090FA80:(mlir::omp::OpenMPDialect::OpenMPDialect(mlir::MLIRContext*))
clang-12: error: linker command failed with exit code 1 (use -v to see invocation)
[17/260] Linking CXX executable bin/opt
ninja: build stopped: subcommand failed.
```

Tried a clean build but to no avail.

**Comment by jeanPerier** [comment]
Should be fixed in fir-dev by ff9e1693c6c1d507c05846023b57345e02348c2e.
Link time dependency was introduced by the pass added in #1061 to support the presence of LLVM IR, OpenMP and OpenMP dialects.

**Comment by oroppas** [comment]
Thanks!
